File: src/main/java/com/merge/api/resources/crm/types/LinkToken.java
package com.merge.api.resources.crm.types;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSetter;
import com.fasterxml.jackson.annotation.Nulls;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import java.util.Objects;
import java.util.Optional;

@JsonInclude(JsonInclude.Include.NON_EMPTY)
@JsonDeserialize(builder = LinkToken.Builder.class)
public final class LinkToken {
    private final String linkToken;

    private final Optional<String> integrationName;

    private final Optional<String> magicLinkUrl;

    private LinkToken(String linkToken, Optional<String> integrationName, Optional<String> magicLinkUrl) {
        this.linkToken = linkToken;
        this.integrationName = integrationName;
        this.magicLinkUrl = magicLinkUrl;
    }

    @JsonProperty("link_token")
    public String getLinkToken() {
        return linkToken;
    }

    @JsonProperty("integration_name")
    public Optional<String> getIntegrationName() {
        return integrationName;
    }

    @JsonProperty("magic_link_url")
    public Optional<String> getMagicLinkUrl() {
        return magicLinkUrl;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) return true;
        return other instanceof LinkToken && equalTo((LinkToken) other);
    }

    private boolean equalTo(LinkToken other) {
        return linkToken.equals(other.linkToken)
                && integrationName.equals(other.integrationName)
                && magicLinkUrl.equals(other.magicLinkUrl);
    }

    @Override
    public int hashCode() {
        return Objects.hash(this.linkToken, this.integrationName, this.magicLinkUrl);
    }

    @Override
    public String toString() {
        return "LinkToken{" + "linkToken: " + linkToken + ", integrationName: " + integrationName + ", magicLinkUrl: "
                + magicLinkUrl + "}";
    }

    public static LinkTokenStage builder() {
        return new Builder();
    }

    public interface LinkTokenStage {
        _FinalStage linkToken(String linkToken);

        Builder from(LinkToken other);
    }

    public interface _FinalStage {
        LinkToken build();

        _FinalStage integrationName(Optional<String> integrationName);

        _FinalStage integrationName(String integrationName);

        _FinalStage magicLinkUrl(Optional<String> magicLinkUrl);

        _FinalStage magicLinkUrl(String magicLinkUrl);
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public static final class Builder implements LinkTokenStage, _FinalStage {
        private String linkToken;

        private Optional<String> magicLinkUrl = Optional.empty();

        private Optional<String> integrationName = Optional.empty();

        private Builder() {}

        @Override
        public Builder from(LinkToken other) {
            linkToken(other.getLinkToken());
            integrationName(other.getIntegrationName());
            magicLinkUrl(other.getMagicLinkUrl());
            return this;
        }

        @Override
        @JsonSetter("link_token")
        public _FinalStage linkToken(String linkToken) {
            this.linkToken = linkToken;
            return this;
        }

        @Override
        public _FinalStage magicLinkUrl(String magicLinkUrl) {
            this.magicLinkUrl = Optional.of(magicLinkUrl);
            return this;
        }

        @Override
        @JsonSetter(value = "magic_link_url", nulls = Nulls.SKIP)
        public _FinalStage magicLinkUrl(Optional<String> magicLinkUrl) {
            this.magicLinkUrl = magicLinkUrl;
            return this;
        }

        @Override
        public _FinalStage integrationName(String integrationName) {
            this.integrationName = Optional.of(integrationName);
            return this;
        }

        @Override
        @JsonSetter(value = "integration_name", nulls = Nulls.SKIP)
        public _FinalStage integrationName(Optional<String> integrationName) {
            this.integrationName = integrationName;
            return this;
        }

        @Override
        public LinkToken build() {
            return new LinkToken(linkToken, integrationName, magicLinkUrl);
        }
    }
}

File: merge-java-client-core/src/main/kotlin/dev/merge/api/models/LinkToken.kt
package dev.merge.api.models

import com.fasterxml.jackson.annotation.JsonAnyGetter
import com.fasterxml.jackson.annotation.JsonAnySetter
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import dev.merge.api.core.ExcludeMissing
import dev.merge.api.core.JsonField
import dev.merge.api.core.JsonMissing
import dev.merge.api.core.JsonValue
import dev.merge.api.core.NoAutoDetect
import dev.merge.api.core.toUnmodifiable
import java.util.Objects
import java.util.Optional

@JsonDeserialize(builder = LinkToken.Builder::class)
@NoAutoDetect
class LinkToken
private constructor(
    private val linkToken: JsonField<String>,
    private val integrationName: JsonField<String>,
    private val magicLinkUrl: JsonField<String>,
    private val additionalProperties: Map<String, JsonValue>,
) {

    private var validated: Boolean = false

    private var hashCode: Int = 0

    fun linkToken(): String = linkToken.getRequired("link_token")

    fun integrationName(): Optional<String> =
        Optional.ofNullable(integrationName.getNullable("integration_name"))

    fun magicLinkUrl(): Optional<String> =
        Optional.ofNullable(magicLinkUrl.getNullable("magic_link_url"))

    @JsonProperty("link_token") @ExcludeMissing fun _linkToken() = linkToken

    @JsonProperty("integration_name") @ExcludeMissing fun _integrationName() = integrationName

    @JsonProperty("magic_link_url") @ExcludeMissing fun _magicLinkUrl() = magicLinkUrl

    @JsonAnyGetter
    @ExcludeMissing
    fun _additionalProperties(): Map<String, JsonValue> = additionalProperties

    fun validate(): LinkToken = apply {
        if (!validated) {
            linkToken()
            integrationName()
            magicLinkUrl()
            validated = true
        }
    }

    fun toBuilder() = Builder().from(this)

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }

        return other is LinkToken &&
            this.linkToken == other.linkToken &&
            this.integrationName == other.integrationName &&
            this.magicLinkUrl == other.magicLinkUrl &&
            this.additionalProperties == other.additionalProperties
    }

    override fun hashCode(): Int {
        if (hashCode == 0) {
            hashCode =
                Objects.hash(
                    linkToken,
                    integrationName,
                    magicLinkUrl,
                    additionalProperties,
                )
        }
        return hashCode
    }

    override fun toString() =
        "LinkToken{linkToken=$linkToken, integrationName=$integrationName, magicLinkUrl=$magicLinkUrl, additionalProperties=$additionalProperties}"

    companion object {

        @JvmStatic fun builder() = Builder()
    }

    class Builder {

        private var linkToken: JsonField<String> = JsonMissing.of()
        private var integrationName: JsonField<String> = JsonMissing.of()
        private var magicLinkUrl: JsonField<String> = JsonMissing.of()
        private var additionalProperties: MutableMap<String, JsonValue> = mutableMapOf()

        @JvmSynthetic
        internal fun from(linkToken: LinkToken) = apply {
            this.linkToken = linkToken.linkToken
            this.integrationName = linkToken.integrationName
            this.magicLinkUrl = linkToken.magicLinkUrl
            additionalProperties(linkToken.additionalProperties)
        }

        fun linkToken(linkToken: String) = linkToken(JsonField.of(linkToken))

        @JsonProperty("link_token")
        @ExcludeMissing
        fun linkToken(linkToken: JsonField<String>) = apply { this.linkToken = linkToken }

        fun integrationName(integrationName: String) =
            integrationName(JsonField.of(integrationName))

        @JsonProperty("integration_name")
        @ExcludeMissing
        fun integrationName(integrationName: JsonField<String>) = apply {
            this.integrationName = integrationName
        }

        fun magicLinkUrl(magicLinkUrl: String) = magicLinkUrl(JsonField.of(magicLinkUrl))

        @JsonProperty("magic_link_url")
        @ExcludeMissing
        fun magicLinkUrl(magicLinkUrl: JsonField<String>) = apply {
            this.magicLinkUrl = magicLinkUrl
        }

        fun additionalProperties(additionalProperties: Map<String, JsonValue>) = apply {
            this.additionalProperties.clear()
            this.additionalProperties.putAll(additionalProperties)
        }

        @JsonAnySetter
        fun putAdditionalProperty(key: String, value: JsonValue) = apply {
            this.additionalProperties.put(key, value)
        }

        fun putAllAdditionalProperties(additionalProperties: Map<String, JsonValue>) = apply {
            this.additionalProperties.putAll(additionalProperties)
        }

        fun build(): LinkToken =
            LinkToken(
                linkToken,
                integrationName,
                magicLinkUrl,
                additionalProperties.toUnmodifiable(),
            )
    }
}
